File: UVa/UVa558_Wormholes.py
import queue

MAX = 2010
INF = int(1e9)

# khai báo class Node với id là tên đỉnh, dist là chi phí
# define operator __lt__ (less than) cho class, so sánh theo dist nhỏ hơn, dùng trong PriorityQueue
class Triad:
    source = 0
    target = 0
    weight = 0
    def __init__(self, source, target, weight):
        self.source = source
        self.target = target
        self.weight = weight

def BellmanFord(source, graph, dist):
    dist[source] = 0
    for i in range(1, n):
        for j in range(m):
            u = graph[j].source
            v = graph[j].target
            w = graph[j].weight
            if dist[u] != INF and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                path[v] = u
    for i in range(m):
        u = graph[i].source
        v = graph[i].target
        w = graph[i].weight
        if dist[u] != INF and dist[u] + w < dist[v]:
            return False
    return True

if __name__ == '__main__':
    t = int(input())
    for cs in range(t):
        n, m = map(int, input().split(' '))
        dist = [INF for i in range(n+5)]
        path = [-1 for i in range(n+5)]
        graph = []
        for i in range(m):
            u, v, w = map(int, input().split(' '))
            graph.append(Triad(u, v, w))
        res = BellmanFord(0, graph, dist)
        if res == False:
            print("possible")
        else:
            print("not possible")

File: UVa/UVa558_Wormholes.java
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Stack;
import java.util.Scanner;

class Triad {
    private int source;
    private int target;
    private int weight;
    public Triad(int source, int target, int weight) {
        this.source = source;
        this.target = target;
        this.weight = weight;
    }
    public int getSource() {
        return this.source;
    } 
    public int getTarget() {
        return this.target;
    }
    public int getWeight() {
        return this.weight;
    }   
}

class FastReader {
    BufferedReader br;
    StringTokenizer st;

    public FastReader()
    {
        br = new BufferedReader(new
                InputStreamReader(System.in));
    }

    String next()
    {
        while (st == null || !st.hasMoreElements())
        {
            try
            {
                st = new StringTokenizer(br.readLine());
            }
            catch (IOException e)
            {
                e.printStackTrace();
            }
        }
        return st.nextToken();
    }

    int nextInt()
    {
        return Integer.parseInt(next());
    }

    long nextLong()
    {
        return Long.parseLong(next());
    }

    double nextDouble()
    {
        return Double.parseDouble(next());
    }

    String nextLine()
    {
        String str = "";
        try
        {
            str = br.readLine();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
        return str;
    }
}

public class Main {    
    public static boolean BellmanFord(int source, ArrayList<Triad> graph, int[] dist) {
        int u, v, w;
        dist[source] = 0;
        for (int j=1; j<=dist.length-1; j++) { // For every node, examine all the edges again
            for (int l=0; l<graph.size(); l++) {
                u = graph.get(l).getSource();
                v = graph.get(l).getTarget();
                w = graph.get(l).getWeight();
                // Dist from source to node u + cost on edge u-v < Dist from source to node v (currently)
                if (dist[u] != Integer.MAX_VALUE && (dist[u] + w < dist[v])) {
                    dist[v] = dist[u] + w;
                    // path[v] = u; // save the path
                }
            }
        }
        // Test if there is a negative cycle
        for (int i=0; i<graph.size(); i++) {
            u = graph.get(i).getSource();
            v = graph.get(i).getTarget();
            w = graph.get(i).getWeight();
            if (dist[u] != Integer.MAX_VALUE && (dist[u] + w < dist[v])) {
                return false;
            } 
        }
        return true;
    }
    
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0) {
            t--;
            int n = sc.nextInt();
            int m = sc.nextInt();
            
            ArrayList<Triad> graph = new ArrayList<Triad>();
            int[] dist = new int[n];
            Arrays.fill(dist, Integer.MAX_VALUE);
            
            for (int i = 0; i < m; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                int w = sc.nextInt();
                graph.add(new Triad(u, v, w));
            }
            if (BellmanFord(0, graph, dist) == false) {
                System.out.println("possible");
            }
            else {
                System.out.println("not possible");
            }
        }
    }
}